packages/btcstaking/src/scripts_utils.rs

```diff
@@ -1,106 +1,96 @@
-    OP_PUSHNUM_1,
 };
 
-use bitcoin::secp256k1::PublicKey;
-use bitcoin::taproot::LeafVersion;
 use bitcoin::ScriptBuf;
-use bitcoin::{TapNodeHash, TapTweakHash, XOnlyPublicKey};
 
-use k256::elliptic_curve::sec1::ToEncodedPoint;
-use k256::elliptic_curve::subtle::Choice;
 use k256::schnorr::VerifyingKey;
-use k256::{
-    elliptic_curve::{ops::MulByGenerator, point::DecompressPoint, PrimeField},
-    AffinePoint, ProjectivePoint, Scalar,
-};
 
-const UNSPENDABLE_KEY: &str = "0250929b74c1a04954b78b4b6035e97a5e078a5a0f28ec96d547bfee9ace803ac0";
+/// assemble_multisig_script is a private helper to assemble multisig script
+/// if `withVerify` is true script will end with OP_NUMEQUALVERIFY otherwise with OP_NUMEQUAL
+/// SCRIPT: <Pk1> OP_CHEKCSIG <Pk2> OP_CHECKSIGADD <Pk3> OP_CHECKSIGADD ... <PkN> OP_CHECKSIGADD <threshold> OP_NUMEQUALVERIFY (or OP_NUMEQUAL)
+fn assemble_multisig_script(
+    pubkeys: &[VerifyingKey],
+    threshold: usize,
+    with_verify: bool,
+) -> Result<ScriptBuf> {
+    let mut builder = Builder::new();
+    for (i, key) in pubkeys.iter().enumerate() {
+        let pk_bytes: [u8; 32] = key.to_bytes().into();
+        builder = builder.push_slice(pk_bytes);
+        if i == 0 {
+            builder = builder.push_opcode(OP_CHECKSIG);
+        } else {
+            builder = builder.push_opcode(OP_CHECKSIGADD);
+        }
+    }
 
-fn unspendable_key_path_internal_pub_key() -> XOnlyPublicKey {
-    let key_bytes = hex::decode(UNSPENDABLE_KEY).unwrap();
+    builder = builder.push_int(threshold as i64);
+    if with_verify {
+        builder = builder.push_opcode(OP_NUMEQUALVERIFY);
+    } else {
+        builder = builder.push_opcode(OP_NUMEQUAL);
+    }
 
-    let (pk_x, _) = PublicKey::from_slice(&key_bytes)
-        .unwrap()
-        .x_only_public_key();
-    pk_x
+    Ok(builder.into_script())
 }
 
-// sort_keys sorts public keys in lexicographical order
-pub fn sort_keys(keys: &mut [VerifyingKey]) {
+/// sort_keys sorts public keys in lexicographical order
+fn sort_keys(keys: &mut [VerifyingKey]) {
     keys.sort_by(|a, b| {
         let a_serialized = a.to_bytes();
         let b_serialized = b.to_bytes();
         a_serialized.cmp(&b_serialized)
     });
 }
 
-/// prepare_keys_for_multisig_script prepares keys for multisig, ensuring there are no duplicates
-pub fn prepare_keys_for_multisig_script(keys: &[VerifyingKey]) -> Result<Vec<VerifyingKey>> {
+/// prepare_keys_for_multisig_script prepares keys to be used in multisig script
+/// Validates whether there are at least 2 keys
+/// and returns copy of the slice of keys sorted lexicographically.
+///
+/// Note: It is up to the caller to ensure that the keys are unique
+fn prepare_keys_for_multisig_script(keys: &[VerifyingKey]) -> Result<Vec<VerifyingKey>> {
     if keys.len() < 2 {
         return Err(Error::InsufficientMultisigKeys {});
     }
 
     let mut sorted_keys = keys.to_vec();
     sort_keys(&mut sorted_keys);
```

Comment:
I would inline the `sort_keys()` function as it's used only once, and both the caller and callee are very short.